### PR TITLE
Support nested structs for `min`/`max` aggregations

### PR DESCRIPTION
**Pull request by ttnghia**
Currently `min` and `max` aggregations can only operate on a single-level structs column. This PR adds full support for arbitrarily nested structs (aka `Struct[Any]` type) for those aggregations.

Depends on:
 * https://github.com/rapidsai/cudf/pull/13069

### CODE REVIEW

**Comment by ttnghia** [comment]
build

File: integration_tests/src/main/python/hash_aggregate_test.py [hunk 1]
```diff
@@ -1833,15 +1833,15 @@ def test_std_variance_partial_replace_fallback(data_gen,
...
+# test min/max aggregations for structs
 #
 gens_for_max_min = [byte_gen, short_gen, int_gen, long_gen,
     float_gen, double_gen,
     string_gen, boolean_gen,
     date_gen, timestamp_gen,
     DecimalGen(precision=12, scale=2),
     DecimalGen(precision=36, scale=5),
-    null_gen]
+    null_gen] + array_gens_sample + struct_gens_sample
```
Comment (revans2) [inline integration_tests/src/main/python/hash_aggregate_test.py:1844]: You enabled UDTs and CalendarInterval, but I see no tests for those. They are types that we don't really support much. UDT we don't support anywhere right now. CalendarInterval we really only support for literals, but there are a few places we can do more. Also Spark does not support MAPs for sorts at all, which we say we support, but the CPU should fail with.  That last one is a nit, but the others we either need to add tests or stop listing that we support them.

Comment (ttnghia) [reply integration_tests/src/main/python/hash_aggregate_test.py:1844]: I've changed the input type check. Please take a look again.

**Comment by ttnghia** [comment]
build

File: integration_tests/src/main/python/hash_aggregate_test.py (hunk 1 above)
Comment (revans2) [reply integration_tests/src/main/python/hash_aggregate_test.py:1844]: nit: We probably could support binary, but I am fine with this for now. Just file a follow on for binary if we don't do it here.

**Review by revans2 — APPROVED** [review]

File: integration_tests/src/main/python/hash_aggregate_test.py (hunk 1 above)
Comment (ttnghia) [reply integration_tests/src/main/python/hash_aggregate_test.py:1844]: I have no idea if cudf can find min/max on binary type. Filing an issue for it.....

Comment (ttnghia) [reply integration_tests/src/main/python/hash_aggregate_test.py:1844]: Filed https://github.com/NVIDIA/spark-rapids/issues/8643.

**Comment by revans2** [comment]
@ttnghia is there a reason that we can support a struct with an array in it, but not an array on it's own?

**Comment by ttnghia** [comment]
cudf can support min/max on any type now. So in the plugin we can just enable array type (with integration test).

**Comment by revans2** [comment]
> cudf can support min/max on any type now. So in the plugin we can just enable array type (with integration test).

I tried that. I get a CUDF error when I do it.

```
ai.rapids.cudf.CudfException: CUDF failure at:/home/jenkins/agent/workspace/jenkins-spark-rapids-jni_nightly-dev-476-cuda11/thirdparty/cudf/cpp/src/groupby/sort/group_single_pass_reduction_util.cuh:108: Unsupported groupby reduction type-agg combination.
	at ai.rapids.cudf.Table.groupByAggregate(Native Method)
```

**Comment by revans2** [comment]
Here is the diff I used to enable it.

```
diff --git a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala
index 709abacfc..03f0a38a7 100644
--- a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala
+++ b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala
@@ -2084,14 +2084,13 @@ object GpuOverrides extends Logging {
       "Max aggregate operator",
       ExprChecksImpl(
         ExprChecks.reductionAndGroupByAgg(
-          (TypeSig.commonCudfTypes + TypeSig.DECIMAL_128 + TypeSig.NULL + TypeSig.STRUCT)
-            .nested(TypeSig.commonCudfTypes + TypeSig.DECIMAL_128 + TypeSig.NULL +
-              TypeSig.STRUCT + TypeSig.ARRAY),
+          (TypeSig.commonCudfTypes + TypeSig.DECIMAL_128 + TypeSig.NULL + TypeSig.STRUCT +
+              TypeSig.ARRAY)
+            .nested(),
           TypeSig.orderable,
           Seq(ParamCheck("input",
-            (TypeSig.commonCudfTypes + TypeSig.DECIMAL_128 + TypeSig.NULL + TypeSig.STRUCT)
-              .nested(TypeSig.commonCudfTypes + TypeSig.DECIMAL_128 + TypeSig.NULL +
-                TypeSig.STRUCT + TypeSig.ARRAY),
+            (TypeSig.commonCudfTypes + TypeSig.DECIMAL_128 + TypeSig.NULL + TypeSig.STRUCT +
+                TypeSig.ARRAY).nested(),
             TypeSig.orderable))).asInstanceOf[ExprChecksImpl].contexts
           ++
           ExprChecks.windowOnly(
@@ -2111,14 +2110,12 @@ object GpuOverrides extends Logging {
       "Min aggregate operator",
       ExprChecksImpl(
         ExprChecks.reductionAndGroupByAgg(
-          (TypeSig.commonCudfTypes + TypeSig.DECIMAL_128 + TypeSig.NULL + TypeSig.STRUCT)
-            .nested(TypeSig.commonCudfTypes + TypeSig.DECIMAL_128 + TypeSig.NULL +
-              TypeSig.STRUCT + TypeSig.ARRAY),
+          (TypeSig.commonCudfTypes + TypeSig.DECIMAL_128 + TypeSig.NULL + TypeSig.STRUCT +
+              TypeSig.ARRAY).nested(),
           TypeSig.orderable,
           Seq(ParamCheck("input",
-            (TypeSig.commonCudfTypes + TypeSig.DECIMAL_128 + TypeSig.NULL + TypeSig.STRUCT)
-              .nested(TypeSig.commonCudfTypes + TypeSig.DECIMAL_128 + TypeSig.NULL +
-                TypeSig.STRUCT + TypeSig.ARRAY),
+            (TypeSig.commonCudfTypes + TypeSig.DECIMAL_128 + TypeSig.NULL + TypeSig.STRUCT +
+                TypeSig.ARRAY).nested(),
             TypeSig.orderable))).asInstanceOf[ExprChecksImpl].contexts
           ++
           ExprChecks.windowOnly(

```

**Comment by ttnghia** [comment]
Oh sorry, I missed that part. More accurately, cudf now supports finding min/max on any types, but it needs a bit more work to enable such functionality. It is similar to doing so for the plugin: we need to enable input type checks and add unit tests.

Please file an issue and I can work on it very soon.

**Comment by revans2** [comment]
I filed https://github.com/rapidsai/cudf/issues/13667 and https://github.com/NVIDIA/spark-rapids/issues/8668 for them

**Pull request merged**